[wubojin] iP

wubojin opened the pull request:
# 🤖 Duke Project Update

This PR includes the completion of several levels of the Duke CLI chatbot, focusing on _enhanced task management_, _error handling_, _date/time support_, and _code quality improvements_.

### Key Updates

- [x] Level 0: Implemented basic commands — `rename`, `greet`, `exit`. 👋 
- [x] Level 1: Introduced the `echo` feature to repeat user input. 🔈 
- [x] Level 2: Added the ability to `add` tasks and `list` all tasks. ➕ 
- [x] Level 3: Enabled tasks to be marked as `done`. ✔️
- [x] Level 4: Developed support for various task types: **ToDos**, **Events**, **Deadlines**. 🗒️ 
- [x] Level 5: Improved error handling for unsupported commands and incorrect input formats. ⚠️
- [x] Level 6: Added the `delete` feature to remove tasks. 🗑️ 
- [x] Level 7: Implemented file saving/loading functionality for persistent task storage. 💾
- [x] Level 8: Enhanced date and time management using `LocalDateTime`. 📅
- [x] Level 9: Added a `find` feature to search tasks by keywords. 🔍 
- [ ] Level 10: Add a GUI to the chatbot using JavaFX. 📱

### Improvements

1. **New Features**:
- Automated text UI testing using `A-TextUiTesting` for robust testing of user commands.
-    Integrated `Enums` for better command management.

2. **Code Quality**:
-    Followed the coding standards and used `CheckStyle` for consistency.
-    Refactored code to enhance **Object-Oriented Programming** principles.
-    Added comprehensive JUnit tests for better code coverage and reliability.


> _"Programming isn't about what you know; it's about what you can figure out."_  
> — Chris Pine ([source](https://www.questglt.com/blogs/50-Best-Quotes-About-Software-Development#:~:text=%22If%20debugging%20is%20the%20process%20of%20removing%20software,code%20that%20humans%20can%20understand.%22%20-%20Martin%20Fowler))

### How to Test

To test these changes, open IntelliJ and run the following command in the terminal:

```shell
./gradlew run

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

wubojin committed:
Level 0: Rename, Greet, Exit

wubojin committed:
Level 1: Echo

wubojin committed:
Level 2: Add, List

wubojin committed:
Level 3: Mark as Done

wubojin committed:
Level 4: ToDos, Events, Deadlines

wubojin committed:
Level 4: ToDos, Events, Deadlines (v1)

wubojin committed:
A-TextUiTesting: Automated Text UI Testing

wubojin committed:
Level 5: Handle Errors

wubojin committed:
Level 6: Delete

wubojin committed:
A-Enums: Use Enums

wubojin committed:
Add Level 7: Save

wubojin committed:
Merge branch 'branch-Level-7'

wubojin committed:
Fix Level 7: Save

wubojin committed:
Merge branch 'branch-Level-7' into branch-Level-8

wubojin committed:
Merge branch 'branch-Level-7'

wubojin committed:
Add Level 8: Dates and Times

wubojin committed:
Merge branch 'branch-Level-8'

wubojin committed:
Add packages

wubojin committed:
Merge remote-tracking branch 'origin/add-gradle-support'

wubojin committed:
Fix packages

wubojin committed:
Add JavaDoc

wubojin committed:
Follow the Coding Standard

wubojin committed:
Add Level 9: Find

wubojin committed:
Fix Level 9: Find

wubojin committed:
Merge parallel branches

niclammm commented on src/main/java/UI/UI.java:
```diff
@@ -0,0 +1,243 @@
+package UI;
```
I think Java Doc comments could be utilise to explain helper methods like displayTaskList, handleMarkCommand eneficial for anyone reading or maintaining the code in the future.

wubojin replied:
Huge oversight on my part, thanks for pointing out!

niclammm commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,92 @@
+                while ((line = reader.readLine()) != null) {
+                    tasks.add(parseTask(line));
+                }
+            } catch (IOException e) {
```
Nice specific error catching

niclammm commented on src/main/java/task/Todo.java:
```diff
@@ -0,0 +1,45 @@
+package task;
+
+/**
+ * Represents a task without any specific date or time.
```
Good use of java doc comments for readability

niclammm commented on src/main/java/task/TaskManager.java:
```diff
@@ -0,0 +1,108 @@
+     *
+     * @param keyword The keyword of the task to be found.
+     */
+    public void findTasks(String keyword) {
```
The findTasks method could be improved by making the search case-insensitive, allowing users to find tasks regardless of the case of the keyword or task description.

wubojin replied:
Great point there, will work on it

niclammm commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,83 @@
+package task;
```
Good use of java doc to explain functions

wubojin committed:
Add more OOP

wubojin committed:
Add JUnit Tests

wubojin committed:
Create a JAR File

wubojin committed:
Fix formatting

wubojin committed:
Use CheckStyle

wubojin committed:
Merge branch 'branch-A-CheckStyle'

meerkatboy commented on src/main/java/task/Todo.java:
```diff
@@ -0,0 +1,45 @@
+     * @param description The description of the todo.
+     * @param isDone Whether the todo is done or not.
+     */
+    public Todo(String description, boolean isDone) {
```
Good use of method overloading to specify different constructors!

meerkatboy commented on src/main/java/ui/Ui.java:
```diff
@@ -0,0 +1,193 @@
+ * Handles user interaction and input for the Jade application.
+ */
+public class Ui {
+    public static final String INDENT = "     "; // 5 spaces for indentation
```
Good that constants are set to screaming snake case!

meerkatboy commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,83 @@
+ * Represents a general task with a description and completion status.
+ */
+public abstract class Task {
+    protected static final DateTimeFormatter INPUT_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
```
good use of access modifiers for your variables!

meerkatboy commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,83 @@
+     *
+     * @return The status of the task as a string.
+     */
+    public String getStatusIcon() {
```
methods are well named in the format of getVariable() which is self explanatory

meerkatboy reviewed (commented):
Overall the coding standards are adhered to! Keep it up!

wubojin committed:
Add Level 10: GUI

wubojin committed:
Add more OOP

wubojin committed:
Merge branch 'master' into branch-Level-10

wubojin committed:
Fix OOP

wubojin committed:
Update Level 10: GUI

wubojin committed:
Fix Level 10: GUI

wubojin committed:
Merge branch 'branch-Level-10'

wubojin committed:
Fix Level 10: GUI

wubojin committed:
Merge branch 'branch-Level-10'

wubojin committed:
Use Varargs and Checkstyle

wubojin committed:
Use Assertions

wubojin committed:
Improve Code Quality

wubojin committed:
Use Streams

wubojin committed:
Merge pull request #1 from wubojin/branch-A-Assertions

Use Assertions to improve error handling

wubojin committed:
Merge master into branch-A-CodeQuality

wubojin committed:
Merge master into branch-A-Streams

wubojin committed:
Merge pull request #2 from wubojin/branch-A-CodeQuality

Improve Code Quality through various refactoring improvements

wubojin committed:
Merge master to branch-A-Streams

wubojin committed:
Merge pull request #3 from wubojin/branch-A-Streams

Use Streams to improve code clarity and efficiency

wubojin committed:
Set up CI

wubojin committed:
Fix GUI

wubojin committed:
Add Sort Extension

wubojin committed:
Merge branch 'branch-C-Sort'

wubojin committed:
Improve Code Quality

Command classes contain duplicate conditional fragments. Empty spaces in messages require manual counting. parse method for GUI-specific cases is not clearly differentiated.

Let's,
- Consolidate duplicated conditional fragments in command classes.
- Use string class repeat method for long empty spaces in messages.
- Rename the parse method for GUI-specific cases to parseForGui for better differentiation and consistency with other method names like runForGui.
- Move commandMap from parse and parseForGui in the Parser class to a private field so future additions of commands will not require modifying both methods.
- Use a constructor in Parser class to accept TaskManager to ensure the access of commandMap. Adjustments were made to Jade class and ParserTest class to accomodate these changes.

wubojin committed:
Fix text-ui-test

wubojin committed:
Better GUI: highlight errors

wubojin committed:
Add User Guide

wubojin committed:
Add JAR file for release

wubojin committed:
Update README.md

wubojin committed:
Update README.md

wubojin committed:
Fix header comments

wubojin committed:
Update READEME.md
